examples/amm/src/contract.rs

```diff
@@ -0,0 +1,276 @@
+                } else {
+                    self.execute_order_remote(&mut result.execution_result, operation)
+                        .await?;
+                }
+            }
+        }
+
+        Ok(result)
+    }
+
+    async fn handle_session_call(
+        &mut self,
+        _context: &CalleeContext,
+        _session: (),
+        _argument: (),
+        _forwarded_sessions: Vec<SessionId>,
+    ) -> Result<SessionCallResult<Self::Message, Self::Response, Self::SessionState>, AmmError>
+    {
+        Err(AmmError::SessionsNotSupported)
+    }
+}
+
+impl Amm {
+    async fn execute_order_local(&mut self, operation: OperationType) -> Result<(), AmmError> {
+        match operation {
+            OperationType::Swap {
+                owner,
+                input_token_idx,
+                output_token_idx,
+                input_amount,
+            } => {
+                if input_token_idx == output_token_idx {
+                    return Err(AmmError::EqualTokensError);
+                }
+
+                let input_pool_balance = self.get_pool_balance(input_token_idx).await?;
+                let output_pool_balance = self.get_pool_balance(output_token_idx).await?;
+
+                let output_amount = amm::calculate_output_amount(
+                    input_amount,
+                    input_pool_balance,
+                    output_pool_balance,
+                )?;
+
+                self.receive_from_account(&owner, input_token_idx, input_amount)
+                    .await?;
+                self.send_to(&owner, output_token_idx, output_amount).await
+            }
+            OperationType::AddLiquidity {
+                owner,
+                token0_amount,
+                token1_amount,
+            } => {
+                let balance0 = self.get_pool_balance(0).await?;
+                let balance1 = self.get_pool_balance(1).await?;
+
+                let total_shares = self.total_shares.get();
+
+                if *total_shares > 0
+                    && abs_diff_ne!(balance0 as f64 / balance1 as f64, token0_amount as f64 / token1_amount as f64, epsilon = 0.01)
+                {
+                    return Err(AmmError::BalanceRatioAlteredError);
+                }
+
+                let new_shares = if *total_shares == 0 {
+                    ((token0_amount * token1_amount) as f64)
+                            .sqrt() as u64
+                } else {
+                    (token0_amount / balance0)
+                        * *total_shares
+                };
+
+                self.receive_from_account(&owner, 0, token0_amount).await?;
+                self.receive_from_account(&owner, 1, token1_amount).await?;
+
+                let total_shares = self.total_shares.get_mut();
+                *total_shares += new_shares;
+
+                Ok(())
+            }
+            OperationType::RemoveLiquidity {
+                owner,
+                shares_amount,
```

Comment:
This means something like: "How many √DOGE √SHIB do you want to remove from the geometric mean"?
I'd find it much more intuitive to just specify the percentage directly, or specify just one of the two tokens, calculating the other accordingly.

Reply:
Oh ok! I think I'll go with the latter